crates/typst/src/math/stretch.rs

```diff
@@ -1,12 +1,156 @@
 use ttf_parser::math::{GlyphAssembly, GlyphConstruction, GlyphPart};
 use ttf_parser::LazyArray16;
 
-use crate::layout::{Abs, Frame, Point, Size};
-use crate::math::{GlyphFragment, MathContext, Scaled, VariantFragment};
+use crate::diag::SourceResult;
+use crate::foundations::{elem, Content, Packed, Resolve, Smart, StyleChain};
+use crate::layout::{Abs, Frame, Length, Point, Rel, Size, VAlignment};
+use crate::math::{
+    GlyphFragment, LayoutMath, MathContext, MathFragment, Scaled, VariantFragment,
+};
 
 /// Maximum number of times extenders can be repeated.
 const MAX_REPEATS: usize = 1024;
 
+/// Stretches a base.
+///
+/// This function can also be used to automatically stretch the base of an
+/// attachment, so that it fits the top and bottom attachments.
+///
+/// Note that only some glyphs can be stretched, and which ones can depends on
+/// the math font being used. Generally however, all math fonts are the same in
+/// this regard.
+///
+/// ```example
+/// $ H stretch(=)^"define" U + p V $
+/// $ f : X stretch(->>, size: #150%)_"surjective" Y $
+/// $ x stretch(harpoons.ltrb, size: #3em) y
+///     stretch(\[, size: #150%) z $
+/// ```
+#[elem(LayoutMath)]
+pub struct StretchElem {
+    /// The base to stretch.
+    #[required]
+    pub base: Content,
+
+    /// The size to stretch to, relative to the glyph's current size.
+    pub size: Smart<Rel<Length>>,
+
+    /// Whether the size should be relative to the width of any top or bottom
+    /// attachments.
+    ///
+    /// ```example
+    /// #set math.stretch(attach: false)
+    /// $ stretch(->, size: #250%)_"very long text" $
+    /// ```
+    #[default(true)]
+    pub attach: bool,
```

Comment:
I think it being always `true` would be consistent with `lr` though:

- `lr((), size: #200%)`: 200% is 200% of parenthesis because body is empty
- `lr((x/2), size: #200%)`: 200% is more than 200% of parenthesis because body is larger

- `stretch(=, size: #200%)`: 200% is 200% of = because no attachment
- `stretch(=, size: #200%)^"hello"`: 200% is 200% of "hello" because it has an attachment

In both cases, it defaults to the width of the stretchable itself if there is nothing else because that's the whole extent of it, but if there's some other grouped thing (body/attachment) it picks it up. In particular, there is no way in `lr` to specify the height relative to just the parenthesis.

Reply:
Yeah okay, that's a fair point. I agree then that it should always be `true`.